File: lib/core/psychrometrics/psychrometrics.dart
import 'dart:math' as math;

const double defaultAtmosphericPressureHPa = 1013.25;

enum HumidityZone {
  dry('trocken'),
  comfortable('angenehm'),
  humid('feucht'),
  critical('kritisch');

  const HumidityZone(this.label);

  final String label;
}

class PsychrometricInput {
  const PsychrometricInput({
    required this.temperatureCelsius,
    required this.relativeHumidityPercent,
    double? pressureHPa,
  }) : pressureHPa = pressureHPa ?? defaultAtmosphericPressureHPa;

  final double temperatureCelsius;
  final double relativeHumidityPercent;
  final double pressureHPa;
}

class PsychrometricResult {
  const PsychrometricResult({
    required this.temperatureCelsius,
    required this.relativeHumidityPercent,
    required this.pressureHPa,
    required this.saturationVaporPressureHPa,
    required this.vaporPressureHPa,
    required this.absoluteHumidityGM3,
    required this.dewPointCelsius,
    required this.zone,
  });

  final double temperatureCelsius;
  final double relativeHumidityPercent;
  final double pressureHPa;
  final double saturationVaporPressureHPa;
  final double vaporPressureHPa;
  final double absoluteHumidityGM3;
  final double? dewPointCelsius;
  final HumidityZone zone;

  double get dewPointSpreadCelsius {
    final dewPoint = dewPointCelsius;
    if (dewPoint == null) {
      return double.infinity;
    }
    return temperatureCelsius - dewPoint;
  }
}

class Psychrometrics {
  const Psychrometrics._();

  static PsychrometricResult calculate(PsychrometricInput input) {
    _validate(input);

    final saturationVaporPressure = saturationVaporPressureHPa(
      input.temperatureCelsius,
    );
    final vaporPressure =
        saturationVaporPressure * input.relativeHumidityPercent / 100;
    final dewPoint = input.relativeHumidityPercent == 0
        ? null
        : dewPointCelsius(
            temperatureCelsius: input.temperatureCelsius,
            relativeHumidityPercent: input.relativeHumidityPercent,
          );
    final absoluteHumidity = absoluteHumidityGM3(
      temperatureCelsius: input.temperatureCelsius,
      vaporPressureHPa: vaporPressure,
    );

    return PsychrometricResult(
      temperatureCelsius: input.temperatureCelsius,
      relativeHumidityPercent: input.relativeHumidityPercent,
      pressureHPa: input.pressureHPa,
      saturationVaporPressureHPa: saturationVaporPressure,
      vaporPressureHPa: vaporPressure,
      absoluteHumidityGM3: absoluteHumidity,
      dewPointCelsius: dewPoint,
      zone: classify(
        temperatureCelsius: input.temperatureCelsius,
        relativeHumidityPercent: input.relativeHumidityPercent,
        dewPointCelsius: dewPoint,
      ),
    );
  }

  static double saturationVaporPressureHPa(double temperatureCelsius) {
    final constants = _magnusConstantsFor(temperatureCelsius);
    return constants.baseHPa *
        math.exp(
          constants.a *
              temperatureCelsius /
              (constants.bCelsius + temperatureCelsius),
        );
  }

  static double dewPointCelsius({
    required double temperatureCelsius,
    required double relativeHumidityPercent,
  }) {
    if (relativeHumidityPercent <= 0) {
      throw ArgumentError.value(
        relativeHumidityPercent,
        'relativeHumidityPercent',
        'must be greater than 0 for dew point calculation',
      );
    }
    if (relativeHumidityPercent > 100) {
      throw ArgumentError.value(
        relativeHumidityPercent,
        'relativeHumidityPercent',
        'must be at most 100',
      );
    }

    final constants = _magnusConstantsFor(temperatureCelsius);
    final gamma =
        math.log(relativeHumidityPercent / 100) +
        constants.a *
            temperatureCelsius /
            (constants.bCelsius + temperatureCelsius);
    return constants.bCelsius * gamma / (constants.a - gamma);
  }

  static double absoluteHumidityGM3({
    required double temperatureCelsius,
    required double vaporPressureHPa,
  }) {
    final kelvin = temperatureCelsius + 273.15;
    return 216.7 * vaporPressureHPa / kelvin;
  }

  static double relativeHumidityForAbsoluteHumidity({
    required double temperatureCelsius,
    required double absoluteHumidityGM3,
  }) {
    final vaporPressureHPa =
        absoluteHumidityGM3 * (temperatureCelsius + 273.15) / 216.7;
    final saturation = saturationVaporPressureHPa(temperatureCelsius);
    return vaporPressureHPa / saturation * 100;
  }

  static HumidityZone classify({
    required double temperatureCelsius,
    required double relativeHumidityPercent,
    required double? dewPointCelsius,
  }) {
    final dewPointSpread = dewPointCelsius == null
        ? double.infinity
        : temperatureCelsius - dewPointCelsius;

    if (relativeHumidityPercent >= 70 || dewPointSpread <= 2) {
      return HumidityZone.critical;
    }
    if (relativeHumidityPercent >= 60) {
      return HumidityZone.humid;
    }
    if (relativeHumidityPercent < 40) {
      return HumidityZone.dry;
    }
    return HumidityZone.comfortable;
  }

  static void _validate(PsychrometricInput input) {
    if (!input.temperatureCelsius.isFinite ||
        input.temperatureCelsius < -80 ||
        input.temperatureCelsius > 80) {
      throw ArgumentError.value(
        input.temperatureCelsius,
        'temperatureCelsius',
        'must be between -80 and 80',
      );
    }
    if (!input.relativeHumidityPercent.isFinite ||
        input.relativeHumidityPercent < 0 ||
        input.relativeHumidityPercent > 100) {
      throw ArgumentError.value(
        input.relativeHumidityPercent,
        'relativeHumidityPercent',
        'must be between 0 and 100',
      );
    }
    if (!input.pressureHPa.isFinite ||
        input.pressureHPa < 300 ||
        input.pressureHPa > 1100) {
      throw ArgumentError.value(
        input.pressureHPa,
        'pressureHPa',
        'must be between 300 and 1100 hPa',
      );
    }
  }

  static _MagnusConstants _magnusConstantsFor(double temperatureCelsius) {
    if (temperatureCelsius < 0) {
      return const _MagnusConstants(a: 22.46, bCelsius: 272.62, baseHPa: 6.112);
    }
    return const _MagnusConstants(a: 17.62, bCelsius: 243.12, baseHPa: 6.112);
  }
}

class _MagnusConstants {
  const _MagnusConstants({
    required this.a,
    required this.bCelsius,
    required this.baseHPa,
  });

  final double a;
  final double bCelsius;
  final double baseHPa;
}

File: backend/app/Support/Psychrometrics.php
<?php

namespace App\Support;

use InvalidArgumentException;

class Psychrometrics
{
    public const DEFAULT_PRESSURE_HPA = 1013.25;

    public static function calculate(
        float $temperatureCelsius,
        float $relativeHumidityPercent,
        ?float $pressureHPa = null,
    ): array {
        $pressureHPa ??= self::DEFAULT_PRESSURE_HPA;
        self::validate($temperatureCelsius, $relativeHumidityPercent, $pressureHPa);

        $saturationVaporPressure = self::saturationVaporPressureHPa($temperatureCelsius);
        $vaporPressure = $saturationVaporPressure * $relativeHumidityPercent / 100;
        $dewPoint = $relativeHumidityPercent === 0.0
            ? null
            : self::dewPointCelsius($temperatureCelsius, $relativeHumidityPercent);
        $absoluteHumidity = self::absoluteHumidityGM3($temperatureCelsius, $vaporPressure);
        $dewPointSpread = $dewPoint === null ? null : $temperatureCelsius - $dewPoint;

        return [
            'temperatureCelsius' => $temperatureCelsius,
            'relativeHumidityPercent' => $relativeHumidityPercent,
            'pressureHPa' => $pressureHPa,
            'saturationVaporPressureHPa' => $saturationVaporPressure,
            'vaporPressureHPa' => $vaporPressure,
            'absoluteHumidityGM3' => $absoluteHumidity,
            'dewPointCelsius' => $dewPoint,
            'dewPointSpreadCelsius' => $dewPointSpread,
            'zone' => self::classify($relativeHumidityPercent, $dewPointSpread),
        ];
    }

    public static function relativeHumidityForAbsoluteHumidity(
        float $temperatureCelsius,
        float $absoluteHumidityGM3,
    ): float {
        $vaporPressureHPa = $absoluteHumidityGM3 * ($temperatureCelsius + 273.15) / 216.7;
        $saturation = self::saturationVaporPressureHPa($temperatureCelsius);

        return $vaporPressureHPa / $saturation * 100;
    }

    public static function saturationVaporPressureHPa(float $temperatureCelsius): float
    {
        $constants = self::magnusConstantsFor($temperatureCelsius);

        return 6.112 * exp(
            $constants['a'] * $temperatureCelsius / ($constants['b'] + $temperatureCelsius)
        );
    }

    private static function dewPointCelsius(
        float $temperatureCelsius,
        float $relativeHumidityPercent,
    ): float {
        $constants = self::magnusConstantsFor($temperatureCelsius);
        $gamma = log($relativeHumidityPercent / 100)
            + $constants['a'] * $temperatureCelsius / ($constants['b'] + $temperatureCelsius);

        return $constants['b'] * $gamma / ($constants['a'] - $gamma);
    }

    private static function absoluteHumidityGM3(
        float $temperatureCelsius,
        float $vaporPressureHPa,
    ): float {
        return 216.7 * $vaporPressureHPa / ($temperatureCelsius + 273.15);
    }

    private static function classify(float $relativeHumidityPercent, ?float $dewPointSpread): string
    {
        if ($relativeHumidityPercent >= 70 || ($dewPointSpread !== null && $dewPointSpread <= 2)) {
            return 'kritisch';
        }

        if ($relativeHumidityPercent >= 60) {
            return 'feucht';
        }

        if ($relativeHumidityPercent < 40) {
            return 'trocken';
        }

        return 'angenehm';
    }

    private static function validate(
        float $temperatureCelsius,
        float $relativeHumidityPercent,
        float $pressureHPa,
    ): void {
        if (! is_finite($temperatureCelsius) || $temperatureCelsius < -80 || $temperatureCelsius > 80) {
            throw new InvalidArgumentException('Temperatur muss zwischen -80 und 80 °C liegen.');
        }

        if (! is_finite($relativeHumidityPercent) || $relativeHumidityPercent < 0 || $relativeHumidityPercent > 100) {
            throw new InvalidArgumentException('Relative Feuchte muss zwischen 0 und 100 % liegen.');
        }

        if (! is_finite($pressureHPa) || $pressureHPa < 300 || $pressureHPa > 1100) {
            throw new InvalidArgumentException('Luftdruck muss zwischen 300 und 1100 hPa liegen.');
        }
    }

    /**
     * @return array{a: float, b: float}
     */
    private static function magnusConstantsFor(float $temperatureCelsius): array
    {
        if ($temperatureCelsius < 0) {
            return ['a' => 22.46, 'b' => 272.62];
        }

        return ['a' => 17.62, 'b' => 243.12];
    }
}
